src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs

```diff
@@ -3774,5 +3774,218 @@ public static void Main()
+System.Console.Write(s.I);
+
+public readonly struct S
+{
+    private readonly int i;
+    public int I { get => i; init => i = value; }
+}
+" }, expectedOutput: "1"); ;
+
+            comp.VerifyIL("<top-level-statements-entry-point>", @"
+{
+  // Code size       31 (0x1f)
+  .maxstack  2
+  .locals init (S V_0, //s
+                S V_1)
+  IL_0000:  ldloca.s   V_1
+  IL_0002:  initobj    ""S""
+  IL_0008:  ldloca.s   V_1
+  IL_000a:  ldc.i4.1
+  IL_000b:  call       ""void S.I.init""
+  IL_0010:  ldloc.1
+  IL_0011:  stloc.0
+  IL_0012:  ldloca.s   V_0
+  IL_0014:  call       ""int S.I.get""
+  IL_0019:  call       ""void System.Console.Write(int)""
+  IL_001e:  ret
+}
+");
+        }
+
+        [Fact]
+        [WorkItem(47612, "https://github.com/dotnet/roslyn/issues/47612")]
+        public void InitOnlyOnReadonlyProperty_AutoProp()
+        {
+            var comp = CompileAndVerify(new[] { IsExternalInitTypeDefinition, @"
+var s = new S { I = 1 };
+System.Console.Write(s.I);
+
+public struct S
+{
+    public readonly int I { get; init; }
+}
+" }, expectedOutput: "1"); ;
+
+            comp.VerifyIL("<top-level-statements-entry-point>", @"
+{
+  // Code size       31 (0x1f)
+  .maxstack  2
+  .locals init (S V_0, //s
+                S V_1)
+  IL_0000:  ldloca.s   V_1
+  IL_0002:  initobj    ""S""
+  IL_0008:  ldloca.s   V_1
+  IL_000a:  ldc.i4.1
+  IL_000b:  call       ""readonly void S.I.init""
+  IL_0010:  ldloc.1
+  IL_0011:  stloc.0
+  IL_0012:  ldloca.s   V_0
+  IL_0014:  call       ""readonly int S.I.get""
+  IL_0019:  call       ""void System.Console.Write(int)""
+  IL_001e:  ret
+}
+");
+        }
+
+        [Fact]
+        [WorkItem(47612, "https://github.com/dotnet/roslyn/issues/47612")]
+        public void InitOnlyOnReadonlyProperty_ManualProp()
+        {
+            var comp = CompileAndVerify(new[] { IsExternalInitTypeDefinition, @"
+var s = new S { I = 1 };
+System.Console.Write(s.I);
+
+public struct S
+{
+    private readonly int i;
+    public readonly int I { get => i; init => i = value; }
+}
+" }, expectedOutput: "1"); ;
+
+            comp.VerifyIL("<top-level-statements-entry-point>", @"
+{
+  // Code size       31 (0x1f)
+  .maxstack  2
+  .locals init (S V_0, //s
+                S V_1)
+  IL_0000:  ldloca.s   V_1
+  IL_0002:  initobj    ""S""
+  IL_0008:  ldloca.s   V_1
+  IL_000a:  ldc.i4.1
+  IL_000b:  call       ""readonly void S.I.init""
+  IL_0010:  ldloc.1
+  IL_0011:  stloc.0
+  IL_0012:  ldloca.s   V_0
+  IL_0014:  call       ""readonly int S.I.get""
+  IL_0019:  call       ""void System.Console.Write(int)""
+  IL_001e:  ret
+}
+");
+        }
+
+        [Fact]
+        [WorkItem(47612, "https://github.com/dotnet/roslyn/issues/47612")]
+        public void InitOnlyOnReadonlyInit_AutoProp()
+        {
+            var comp = CompileAndVerify(new[] { IsExternalInitTypeDefinition, @"
+var s = new S { I = 1 };
+System.Console.Write(s.I);
+
+public struct S
+{
+    public int I { get; readonly init; }
```

Comment:
I think it is strange to allow this, and I also think it is strange to consider an 'init' accessor to be readonly--including when the containing type is 'readonly'. It's effectively saying: "The 'this' parameter to the accessor is a readonly ref, and the accessor is allowed to write to it". It can't cause any differences in behavior.

Also, due to the implementation of `MethodSymbol.IsEffectivelyReadOnly`, the public API will say that `IMethodSymbol.IsReadOnly` is `false` for the  `readonly init;` accessor here. We should add a test for this accessor and a test for when the containing type is `readonly`.

The more I look at this, the stranger I find it--but if you are sure that you do not want to introduce a restriction here, I won't belabor the point.

Reply:
I just think there's no real point to introduce a restriction here. This PR makes it so you can put an `init` in a `readonly struct`, and this is legal because the initter can only be called during the construction phase, when instance readonly is allowed to be violated. Doing this makes the language more regular. It might be worth clarifying with LDM as to whether we should allow this, but I just don't think we get any real benefits to restricting it.

Reply:
I agree it's a bit strange at first glance but I agree that it makes the language more regular to allow it. The mental model of `readonly struct` is that it's a `struct` where every single member has the `readonly` modifier. Hence if we allow `init` in a `readonly struct` then it should follow that we allow `readonly init` in a standard `struct`.

Reply:
In a readonly struct, *most* of the *instance* members are readonly. The constructors are not readonly. The static members are not readonly. "All the members have 'readonly'" is a reasonable mental model but it is a simplification of the reality of the language.